File: tools/mwcc.py
#!/usr/bin/env python3


import os
import pathlib
import subprocess
import sys
import tempfile


os.environ['WINEDEBUG'] = '-all'

argv = sys.argv[1:]
for i in range(len(sys.argv) - 1):
    if argv[i] in ['-Mfile', '-MMfile', '-MDfile', '-MMDfile']:
        makefile_path = pathlib.Path(argv[i + 1])
        break

with tempfile.TemporaryFile() as stdout:
    with tempfile.TemporaryFile() as stderr:
        returncode = subprocess.run(
            ['wine', os.path.join('tools', 'cw', 'modified_mwcceppc.exe'), *argv],
            stdout = stdout,
            stderr = stderr,
        ).returncode
        stdout.seek(0)
        sys.stdout.buffer.write(stdout.read())
        sys.stdout.buffer.flush()
        stderr.seek(0)
        sys.stdout.buffer.write(stderr.read())
        sys.stdout.buffer.flush()
        if returncode != 0:
            sys.exit(returncode)

if makefile_path is not None:
    makefile_in = makefile_path.read_text()
    makefile_out = ''
    while makefile_in != '':
        if makefile_in.startswith('Z:'):
            path = makefile_in.split(None, 1)[0]
            makefile_out += subprocess.check_output(
                ['winepath', '-u', path],
                encoding = 'utf-8',
            ).rstrip('\n')
            makefile_in = makefile_in.split(path, 1)[1]
        else:
            makefile_out += makefile_in[0]
            makefile_in = makefile_in[1:]
    makefile_path.write_text(makefile_out)

File: tools/src/bin/mwcc.rs
use std::env;
use std::fs;
use std::process::Command;

use anyhow::{Context, Result};

fn main() -> Result<()> {
    let args: Vec<_> = env::args().skip(1).collect();
    let makefile_path = args
        .array_windows()
        .find(|[key, _]| matches!(key.as_str(), "-Mfile" | "-MMfile" | "-MDfile" | "-MMDfile"))
        .map(|[_, value]| value)
        .cloned();

    let exit_status = Command::new("wine")
        .arg("tools/cw/modified_mwcceppc.exe")
        .args(args)
        .env("WINEDEBUG", "-all")
        .spawn()?
        .wait()?;
    anyhow::ensure!(exit_status.success());

    let Some(makefile_path) = makefile_path else { return Ok(()) };
    let makefile = fs::read_to_string(&makefile_path)?;
    let current_dir = env::current_dir()?;
    let current_dir = current_dir.to_str().context("invalid UTF-8")?;
    let prefix = format!("Z:{}\\", current_dir.replace('/', "\\"));
    let makefile =
        makefile.replace(&prefix, "").replace('\\', "/").replace('\r', "").replace("/\n", "\\\n");
    fs::write(makefile_path, makefile)?;
    Ok(())
}
